### ISSUE TITLE
[Bug]: fabricCanvasInstance.off("event:name") shows wrong parameter error

### ISSUE THREAD

**Issue by AlvesJorge** [fabricjs/fabric.js#9972]
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have read and followed the [Issue Tracker Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md#%EF%B8%8F-issue-tracker)
- [X] I have searched and referenced existing issues and discussions
- [X] I am filing a **BUG** report.
- [X] I have managed to reproduce the bug **after** upgrading to the **latest** version
- [X] I have created an accurate and minimal reproduction

### Version

6.0.1

### In What environments are you experiencing the problem?

Chrome

### Node Version (if applicable)

None

### Link To Reproduction
[codesandbox](https://codesandbox.io/p/devbox/fabric-nextjs-sandbox-forked-yhcpzs?workspaceId=b726377e-674d-40a3-af17-5ad481d37eb1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cly79ft5h00083b6l5midst8j%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cly79ft5g00033b6l6y9qkf5o%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cly79ft5g00053b6l9u4h8l2e%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cly79ft5g00073b6lhaoymgt5%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cly79ft5g00033b6l6y9qkf5o%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly79ft5g00023b6l0wc6awwj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cly79gn65004i3b6l8n33kdqd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A17%252C%2522startColumn%2522%253A32%252C%2522endLineNumber%2522%253A17%252C%2522endColumn%2522%253A32%257D%255D%252C%2522filepath%2522%253A%2522%252Fpages%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cly79ft5g00033b6l6y9qkf5o%2522%252C%2522activeTabId%2522%253A%2522cly79gn65004i3b6l8n33kdqd%2522%257D%252C%2522cly79ft5g00073b6lhaoymgt5%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly79ft5g00063b6l5m2c5j34%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cly79ft5g00073b6lhaoymgt5%2522%252C%2522activeTabId%2522%253A%2522cly79ft5g00063b6l5m2c5j34%2522%257D%252C%2522cly79ft5g00053b6l9u4h8l2e%2522%253A%257B%2522id%2522%253A%2522cly79ft5g00053b6l9u4h8l2e%2522%252C%2522activeTabId%2522%253A%2522cly79j8gm003d3b6jt0khj0m2%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly79ft5g00043b6l6eyp1d0w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522cly79j8gm003d3b6jt0khj0m2%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cly79j8kf003xdjfs3qcob0iq%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps To Reproduce

Open the sandbox link and notice the hint for the canvas.off("object:added").
Notice how it shows that string isnt compatible with the parameter.

Locally in my private project I get this error hint, but the method works as expected, meaning, the event listener is turned off.   

### Expected Behavior

No hint saying `string` is not a valid parameter.

### Actual Behavior

Hint saying `string` is not a valid parameter.

### Error Message & Stack Trace

```bash
Type '"object:added"' has no properties in common with type 'EventRegistryObject<CanvasEvents>'.typescript(2559)
```

**Comment by AlvesJorge** [comment]
Similarly:
For the objects returned from `fabric.util.enlivenObjects`, when I try to `canvas.add(object)` I get this type error.
```
Argument of type 'TFiller | FabricObject<...> | Shadow | BaseFilter<...> | FabricObject<...>' is not assignable to parameter of type 'FabricObject<Partial<FabricObjectProps>, SerializedObjectProps, ObjectEvents>'.
  Type 'Gradient<"linear", "linear">' is missing the following properties from type 'FabricObject<Partial<FabricObjectProps>, SerializedObjectProps, ObjectEvents>': noScaleCache, lockMovementX, lockMovementY, lockRotation, and 218 more.ts(2345)
```

Full code:
```js
fabric.util.enlivenObjects(myObjects)
  .then(objs => {
    objs.forEach(obj => fabricCanvas.add(obj));
  });
```

**Comment by AlvesJorge** [comment]
And BTW Huge congrats on 6 :) Really happy for you

**Comment by asturur** [comment]
@AlvesJorge let's keep the 2 issues on separate threads please, so i can track fixing one by one.
CAn you open an issue describing the second one?

**Comment by asturur** [comment]
I see the issue in the types.
Also please keep in mind this pattern is a bit unsuggested.
I will fix the type but deprecate the pattern anyway.
The issue with that is that doing off('something') you ma kill some internal fabric functionality based on events.

**Comment by AlvesJorge** [comment]
Yep I will create the second issue, I only commented it here because perhaps they would both not have a solution or not be problems in your eyes.

**Comment by AlvesJorge** [comment]
> Also please keep in mind this pattern is a bit unsuggested.

Just so you have an idea of my use-case:
I keep a history of objects added and removed to my Canvas, and then on a users "Undo" I restore the previous group of objects to the canvas.
To keep track of these History States, I use `object:added` and `object:modified` to listen to changes and create a new entry in the history array.

But before I re-add them, I need to disable the event that updates my history, or else the re-adding of the objects in the "undo" action will trigger the `object:added` creating a new entry in the history :)

**Comment by asturur** [comment]
Well how many event handlers you have for object:added? one? you have 3 solutions AT LEAT:
- use the off with the handler, `canvas.off('object:added', myHandler)`
- use an events handler that can be paused by looking at some value on the canvas ( canvas.historyTracking ) so that when you are restoring history you disable tracking and when you are done you enable it.
- **do NOT use events for code that is not run by user actions**. You probably wrote a function to add an object to the canvas, and you attacched to some button or other type of UI widget. Track your history there. is more manageable in my opinion. Have an addObjectWithHistory function that you use on the UI that also update history and when restoring history you a simpler function that does not have the history lines of code.

Then every application has its strong and weak points, but in general using those events that can fire for many reasons to build flows is not a great idea.

**Comment by asturur** [comment]
To be clear, object:added and object:removed where here before i joined fabricJS, but i would never added them.
I would have added only user input device events, high level actions ( moving, transforming ) and selection events. Things that are run entirely by the developer.

**Comment by AlvesJorge** [comment]
Those are some great suggestions, I apreciate you going out of your way to help me with an issue not directly related to your library!

**Comment by asturur** [comment]
```ts
  /**
   * Unsubscribe all event listeners for eventname.
   * Do not use this pattern. You could kill internal fabricJS events.
   * We know we should have protected events for internal flows, but we don't have yet
   * @deprecated
   * @param {string} eventName event name (eg. 'after:render')
   */
  off<K extends keyof EventSpec>(eventName: K): void;
  /**
   * unsubscribe an event listener
   * @param {string} eventName event name (eg. 'after:render')
   * @param {TEventCallback} handler event listener to unsubscribe
   */
  off<K extends keyof EventSpec>(eventName: K, handler: TEventCallback): void;
```

with type definition here it is usable and is clear that we do not think is a great idea to do a generic off.
Not sure of jsdocs specs, but vscode is smart enough to show you the correct definition depending on the usage

**Comment by asturur** [comment]
closed via #9978
